catalog/plugin/catalog-plugin-defaultsecurityattributevalues/src/main/java/org/codice/alliance/catalog/plugin/defaultsecurity/DefaultSecurityAttributeValuesPlugin.java

```diff
@@ -26,157 +33,175 @@
       return metacard;
     }
 
     if (isRegistryMetacard(metacard)) {
       return metacard;
     }
 
-    final Metacard extendedMetacard;
-    MetacardImpl metacardImpl;
+    final MetacardImpl extendedMetacard;
+    // If the metacard doesn't have any of the securityAttributes AttributeDescriptors, add ALL of
+    // the securityAttributes AttributeDescriptors to the metacard.
     if (securityAttributes
         .getAttributeDescriptors()
         .stream()
         .noneMatch(ad -> metacard.getMetacardType().getAttributeDescriptors().contains(ad))) {
-      metacardImpl =
+      extendedMetacard =
           new MetacardImpl(
               metacard,
               new MetacardTypeImpl(
                   metacard.getMetacardType().getName(),
                   metacard.getMetacardType(),
                   securityAttributes.getAttributeDescriptors()));
     } else {
-      metacardImpl = new MetacardImpl(metacard);
+      extendedMetacard = new MetacardImpl(metacard);
     }
-    Set<String> updatedTags = new HashSet<>(metacard.getTags());
-    updatedTags.add(DEFAULTMARKINGS);
-    metacardImpl.setTags(updatedTags);
-    extendedMetacard = metacardImpl;
 
-    securityMarkings
-        .keySet()
-        .stream()
-        .filter(
-            securityMarking ->
-                extendedMetacard.getMetacardType().getAttributeDescriptor(securityMarking) != null)
-        .forEach(
-            securityMarking -> {
-              extendedMetacard.setAttribute(securityMarkings.get(securityMarking));
-            });
-    return extendedMetacard;
-  }
-
-  public void setClassification(String classification) {
-    addMapping(classification, Security.CLASSIFICATION);
-  }
-
-  public void setReleasability(String releasability) {
-    addMapping(releasability, Security.RELEASABILITY);
-  }
-
-  public void setCodewords(String codewords) {
-    addMapping(codewords, Security.CODEWORDS);
-  }
-
-  public void setDisseminationControls(String disseminationControls) {
-    addMapping(disseminationControls, Security.DISSEMINATION_CONTROLS);
-  }
+    final Set<String> updatedTags = new HashSet<>(metacard.getTags());
+
+    for (Map.Entry<String, List<String>> entry :
+        cachedMetacardAttributeNameToSystemHighValuesMap.entrySet()) {
+      String metacardAttributeName = entry.getKey();
+      List<String> systemHighAttributeValues = entry.getValue();
+      if (extendedMetacard.getMetacardType().getAttributeDescriptor(metacardAttributeName)
+          != null) {
+        final List<Serializable> newAttributeValues =
+            systemHighAttributeValues
+                .stream()
+                .map(Serializable.class::cast)
+                .collect(Collectors.toList());
+        final Attribute newAttribute = new AttributeImpl(metacardAttributeName, newAttributeValues);
+        extendedMetacard.setAttribute(newAttribute);
+        updatedTags.add(DEFAULT_MARKINGS_TAG);
```

Comment:
I know updatedTags is a set but do we want to add this every time? Seems like if we get to this loop we know we want to add the default markings tag

Reply:
It's possible that the metacard could go through the loop without having any attributes set, for example, when none of the system high attribute names in the configuration are found in the `system.attributes` file.